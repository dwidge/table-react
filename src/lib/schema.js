import React from 'react'
import { onChange, onChangeChecks } from '@dwidge/lib-react'
import { getItemById } from '@dwidge/lib'

import { dateYYMMDDfromSeconds, dateSecondsFromYYMMDD } from './date.js'
import isMatch from 'date-fns/isMatch'

export const ColumnText = (name) => ({
	name,
	row(value) {
		return (<column-text key={name}>{value}</column-text>)
	},
	edit(value, setvalue) {
		return (
			<column-text key={name}>
				<input data-testid={'input' + name} value={value} onChange={onChange(setvalue)} />
			</column-text>
		)
	},
	cleanup(value) {
		return value
	},
})

export const ColumnDate = (name) => ({
	name,
	_load(value) {
		return dateYYMMDDfromSeconds(value)
	},
	_save(value) {
		return dateSecondsFromYYMMDD(value)
	},
	valid(value) {
		return isMatch(value, 'yyyy/MM/dd')
	},
	row(value) {
		return (<column-text key={name}>{(value)}</column-text>)
	},
	edit(value, setvalue) {
		return (
			<column-text key={name}>
				<input data-testid={'input' + name} style={this.valid(value) ? {} : { background: 'red' }} value={(value)} onChange={onChange(v => setvalue((v)))} />
			</column-text>
		)
	},
	cleanup(value) {
		return value
	},
})

export const ColumnSet = (name, all, toString) => ({
	name,
	row(value) {
		return (<column-set key={name}>{value.length
			? value.map(id => getItemById(all, id)).map(entry =>
				(<div key={entry.id}>{toString(entry)}</div>),
			)
			: '-'}</column-set>)
	},
	edit(value, setvalue) {
		return (<column-set key={name}>{all.map(entry => (id =>
			(<div key={id}><input data-testid={'input' + name + id} type="checkbox" checked={value.includes(id)} onChange={onChangeChecks(id, value, setvalue)} /> {toString(entry)}</div>))(entry.id),
		)}</column-set>)
	},
	cleanup(value) {
		return value.filter(id => getItemById(all, id))
	},
})
