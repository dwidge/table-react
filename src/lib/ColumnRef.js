import React, { useState, useEffect } from 'react'
import Form from 'react-bootstrap/Form'
import { Typeahead } from 'react-bootstrap-typeahead'

export const getItemBy = (a, v, k = 'id') =>
	a.find(o => o[k] === v)

const se = (x, f = '', t = '' + x) => x == null || x === '' ? f : t

export const ColumnRef = (name, { all, colRef = 'id', colView = 'name', colDisplay = colView }) => ({
	name,
	valid(value) {
		return !value || !!this.lookup(value)
	},
	row(value) {
		const display = this.lookup(value, colDisplay)
		return (<column-text key={name}>
			{se(display, '-')}
		</column-text>)
	},
	sort(a, b) {
		const astr = this.lookup(a, colDisplay)
		const bstr = this.lookup(b, colDisplay)
		return ('' + astr).localeCompare('' + bstr)
	},
	edit(ref, setref) {
		const [view, setview] = useState(se(this.lookup(ref)))

		useEffect(() => {
			setview(se(this.lookup(ref)))
		}, [all])

		const onref = v => {
			setref(v)
			setview(se(this.lookup(v)))
		}
		const onview = v => {
			setview(v)
			setref(se(this.rlookup(v)))
		}

		const [refTa, refTaSet] = [se(ref, [], ['' + ref]), ([s]) => onref(se(s))]
		const refOptions = all.map(v => se(v[colRef]))

		const [viewTa, viewTaSet] = [se(view, [], ['' + view]), ([s]) => onview(se(s))]

		const viewOptions = all.map(v => se(v[colView]))

		return (<column-text key={name}>
			<Form.Group>
				<Form.Label>{colRef}</Form.Label>
				<Typeahead
					id={'input' + name}
					inputProps={{
						'data-testid': 'input' + name,
					}}
					style={this.lookup(ref) ? { minWidth: '10em' } : { minWidth: '10em', borderColor: 'red' }}
					onChange={refTaSet}
					options={refOptions}
					placeholder="Choose..."
					selected={refTa}
				/>
			</Form.Group>
			<Form.Group>
				<Form.Label>{colView}</Form.Label>
				<Typeahead
					id={'inputView' + name}
					inputProps={{
						'data-testid': 'inputView' + name,
					}}
					style={this.rlookup(view) ? { minWidth: '10em' } : { minWidth: '10em', borderColor: 'red' }}
					onChange={viewTaSet}
					options={viewOptions}
					placeholder="Choose..."
					selected={viewTa}
				/>
			</Form.Group>
		</column-text>)
	},
	cleanup(value) {
		return value
	},
	lookup(value, col = colView) {
		const item = getItemBy(all, value, colRef) || getItemBy(all, +value, colRef) || {}
		return item[col]
	},
	rlookup(value) {
		const item = getItemBy(all, value, colView) || getItemBy(all, +value, colView) || {}
		return item[colRef]
	},
})
