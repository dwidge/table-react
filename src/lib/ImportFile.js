import React, { useState, useEffect } from 'react'
import PropTypes from 'prop-types'
import { useFilePicker } from 'use-file-picker'
import Button from 'react-bootstrap/Button'

export function ImportFile({ ext, onAccept, preview = true }) {
	const [openFileSelector, { filesContent, loading }] = useFilePicker({
		accept: ext,
		multiple: false,
		limitFilesConfig: { max: 1 },
	})
	const [file, setfile] = useState()

	useEffect(() => {
		setfile(filesContent[0])
	}, [filesContent])

	return (
		<div>
			<Button data-testid={'buttonImport' + ext} onClick={() => openFileSelector()}>Import {ext}</Button>
			{loading
				? (<div>Loading...</div>)
				: file
					? (
						<div style={{ border: 'solid 1px orange', margin: '.5em' }}>
							<Button onClick={() => { onAccept(file.content); setfile() }}>Accept</Button>
							<h4>{file.name}</h4>
							{preview ? (<pre style={{ overflow: 'scroll', whiteSpace: 'pre-wrap' }}>{file.content}</pre>) : ''}
						</div>)
					: ''}
		</div>
	)
}

ImportFile.propTypes = {
	ext: PropTypes.string,
	onAccept: PropTypes.func.isRequired,
	preview: PropTypes.bool,
}
