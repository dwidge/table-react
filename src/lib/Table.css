table-table {
display: table;
width: 100%;
margin-top: 1em;
}
table-header, table-item {
display: table-row;
}
table-header {
background: orange;
}
table-header > *, table-item > * {
display: table-cell;
vertical-align: middle;
padding: 0.5em;
}
table-item > * > input {
width: 90%;
}
table-buttons {
width: 20%;
padding: 0;
}
table-buttons > button {
width: 100%;
}
