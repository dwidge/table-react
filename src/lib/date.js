export const dateYYMMDDfromSeconds = sec =>
	new Date(sec * 1000).toLocaleDateString('en-ZA').split('/').join('-')

export const dateSecondsFromYYMMDD = str => {
	const segs = str.split('-')
	const date = new Date(+segs[0], (+segs[1]) - 1, +segs[2])
	return date.getTime() / 1000 | 0
}
