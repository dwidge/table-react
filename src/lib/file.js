import { saveAs } from 'file-saver'

export const saveText = (text, file, mime = 'text/plain') => {
	const blob = new Blob([text], { type: mime + ';charset=utf-8' })
	saveAs(blob, file)
}
