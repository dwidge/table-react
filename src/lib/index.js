export { ColumnText, ColumnSet, ColumnDate } from './schema'
export { Table } from './Table'
