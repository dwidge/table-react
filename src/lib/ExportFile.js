import React, { useState } from 'react'
import PropTypes from 'prop-types'
import { saveText } from './file'
import Button from 'react-bootstrap/Button'

export function ExportFile({ name, ext, content, preview = true }) {
	const [file, setfile] = useState()

	return (
		<div>
			<Button data-testid={'buttonExport' + ext} onClick={() => setfile({ name, content })}>Export {ext}</Button>
			{file
				? (
					<div style={{ border: 'solid 1px orange', margin: '.5em' }}>
						<Button data-testid={'buttonExportAccept' + ext} onClick={() => { saveText(file.content, file.name); setfile() }}>Accept</Button>
						<h4>{file.name}</h4>
						{preview ? (<pre style={{ overflow: 'scroll', whiteSpace: 'pre-wrap' }}>{file.content}</pre>) : ''}
					</div>)
				: ''}
		</div>
	)
}

ExportFile.propTypes = {
	name: PropTypes.string.isRequired,
	ext: PropTypes.string.isRequired,
	content: PropTypes.string.isRequired,
	preview: PropTypes.bool,
}
