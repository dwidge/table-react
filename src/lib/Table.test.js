import React, { useState } from 'react'
import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'

import { Table, ColumnText, ColumnSet } from '.'
import App from './../App'

import * as J from '@dwidge/lib-react'
import * as Lib from '@dwidge/lib'
const text = J.tools(userEvent, screen, jest).text
const type = J.type(userEvent, screen)
const clear = J.clear(userEvent, screen)
const click = J.click(userEvent, screen)
const serialSpy = J.serialSpy(jest)
const input = async (id, text) => {
	await clear(id)
	await type(id, text)
}

jest.mock('@dwidge/lib', () => {
	return {
		__esModule: true,
		...jest.requireActual('@dwidge/lib'),
	}
})
beforeEach(async () => {
	serialSpy(Lib, 'uuid', [1, 2, 3])
})
afterEach(() => {
	jest.restoreAllMocks()
})

describe('Table', () => {
	const b1 = { id: 1, bc: 'c1' }
	const a0 = () => ({ id: Lib.uuid(), aa: 'a', ab: [], ac: '2000/01/01', ad: undefined })
	const a1 = { id: 1, aa: 'a1', ab: [1], ac: '2001/01/01' }
	const a2 = { id: 2, aa: 'a2', ab: [], ac: '2002/01/01', ad: 1 }

	it('enters row into list', async () => {
		Lib.uuid()
		Lib.uuid()
		render(<App/>)
		await click('buttonAdd')
		expect(screen.getByTestId('tableA')).toMatchSnapshot()
		await input('inputColA', 'a3')
		await input('inputColC', '2022-01-04')
		await click('buttonSave')
		expect(screen.getByTestId('tableA')).toMatchSnapshot()
		await input('inputColC', '2022/01/04')
		await input('inputColD', '1')
		await click('buttonSave')
		expect(screen.getByTestId('tableA')).toMatchSnapshot()
		expect(screen.getByTestId('msg')).toMatchSnapshot()
		await click('buttonColE3')
		expect(screen.getByTestId('msg')).toMatchSnapshot()
	})

	it('enables a slot', async () => {
		const Frag = () => (<Table name='A' schema={{
			aa: ColumnText('ColA'),
			ab: ColumnSet('ColB', [b1], val => val.bc),
		}} newRow={a0} rows={useState([])} />)
		render(<Frag/>)
		await click('buttonAdd')
		await input('inputColA', 'a1')
		await click('inputColB1')
		await click('buttonSave')
		expect(screen.getByTestId('tableA')).toMatchSnapshot()
	})

	it('clears missing slot', async () => {
		const Frag = () => (<Table name='A' schema={{
			aa: ColumnText('ColA'),
			ab: ColumnSet('ColB', [], val => val.bc),
		}} newRow={a0} rows={useState([a1])} />)
		render(<Frag/>)
		expect(screen.getByTestId('tableA')).toMatchSnapshot()
	})

	it('confirms & clears list', async () => {
		const Frag = () => (<Table name='A' schema={{
			aa: ColumnText('ColA'),
			ab: ColumnSet('ColB', [b1], val => val.bc),
		}} newRow={a0} rows={useState([a1, a2])} />)
		render(<Frag/>)
		expect(screen.getByTestId('tableA')).toMatchSnapshot()
		expect(text('buttonClear')).toEqual('Clear')
		await click('buttonClear')
		expect(text('buttonClear')).toEqual('Confirm')
		expect(screen.getByTestId('tableA')).toMatchSnapshot()
		await click('buttonClear')
		expect(text('buttonClear')).toEqual('Clear')
		expect(screen.getByTestId('tableA')).toMatchSnapshot()
	})

	it('addDel', async () => {
		const Frag = () => (<Table name='A' schema={{
			aa: ColumnText('ColA'),
			ab: ColumnSet('ColB', [b1], val => val.bc),
		}} newRow={a0} rows={useState([a1, a2])} addDel={false} />)
		render(<Frag/>)
		expect(screen.getByTestId('tableA')).toMatchSnapshot()
	})

	it('pages', async () => {
		const Frag = () => (<Table name='A' schema={{
			aa: ColumnText('ColA'),
			ab: ColumnSet('ColB', [b1], val => val.bc),
		}} newRow={a0} rows={useState([a1, a2])} pageLength={1} />)
		render(<Frag/>)
		expect(screen.getByTestId('tableA')).toMatchSnapshot()
	})

	it('pages next prev', async () => {
		const Frag = () => (<Table name='A' schema={{
			aa: ColumnText('ColA'),
			ab: ColumnSet('ColB', [b1], val => val.bc),
		}} newRow={a0} rows={useState([a1, a2])} pageLength={1} />)
		render(<Frag/>)
		await click('buttonPrev')
		expect(screen.getByTestId('tableA')).toMatchSnapshot()
		await click('buttonNext')
		expect(screen.getByTestId('tableA')).toMatchSnapshot()
		await click('buttonNext')
		expect(screen.getByTestId('tableA')).toMatchSnapshot()
		await click('buttonPrev')
		expect(screen.getByTestId('tableA')).toMatchSnapshot()
	})

	it('inlineHeaders', async () => {
		const Frag = () => (<Table name='A' schema={{
			aa: ColumnText('ColA'),
			ab: ColumnSet('ColB', [b1], val => val.bc),
		}} newRow={a0} rows={useState([a1, a2])} pageLength={1} inlineHeaders={true} />)
		render(<Frag/>)
		expect(screen.getByTestId('tableA')).toMatchSnapshot()
	})
})
