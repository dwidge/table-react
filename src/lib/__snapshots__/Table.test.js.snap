// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table addDel 1`] = `
<div-page
  data-testid="tableA"
>
  <table-table>
    <table-header>
      <column-header>
        ColA
      </column-header>
      <column-header>
        ColB
      </column-header>
    </table-header>
    <table-item>
      <column-text>
        a1
      </column-text>
      <column-set>
        <div>
          c1
        </div>
      </column-set>
      <table-buttons>
        <button
          data-testid="buttonEdit1"
        >
          Edit
        </button>
      </table-buttons>
    </table-item>
    <table-item>
      <column-text>
        a2
      </column-text>
      <column-set>
        -
      </column-set>
      <table-buttons>
        <button
          data-testid="buttonEdit2"
        >
          Edit
        </button>
      </table-buttons>
    </table-item>
  </table-table>
</div-page>
`;

exports[`Table clears missing slot 1`] = `
<div-page
  data-testid="tableA"
>
  <table-table>
    <table-header>
      <column-header>
        ColA
      </column-header>
      <column-header>
        ColB
      </column-header>
    </table-header>
    <table-item>
      <column-text>
        a1
      </column-text>
      <column-set>
        -
      </column-set>
      <table-buttons>
        <button
          data-testid="buttonEdit1"
        >
          Edit
        </button>
        <button
          data-testid="buttonDel1"
        >
          Del
        </button>
      </table-buttons>
    </table-item>
  </table-table>
  <button
    data-testid="buttonAdd"
  >
    Add
  </button>
  <button
    data-testid="buttonClear"
  >
    Clear
  </button>
</div-page>
`;

exports[`Table confirms & clears list 1`] = `
<div-page
  data-testid="tableA"
>
  <table-table>
    <table-header>
      <column-header>
        ColA
      </column-header>
      <column-header>
        ColB
      </column-header>
    </table-header>
    <table-item>
      <column-text>
        a1
      </column-text>
      <column-set>
        <div>
          c1
        </div>
      </column-set>
      <table-buttons>
        <button
          data-testid="buttonEdit1"
        >
          Edit
        </button>
        <button
          data-testid="buttonDel1"
        >
          Del
        </button>
      </table-buttons>
    </table-item>
    <table-item>
      <column-text>
        a2
      </column-text>
      <column-set>
        -
      </column-set>
      <table-buttons>
        <button
          data-testid="buttonEdit2"
        >
          Edit
        </button>
        <button
          data-testid="buttonDel2"
        >
          Del
        </button>
      </table-buttons>
    </table-item>
  </table-table>
  <button
    data-testid="buttonAdd"
  >
    Add
  </button>
  <button
    data-testid="buttonClear"
  >
    Clear
  </button>
</div-page>
`;

exports[`Table confirms & clears list 2`] = `
<div-page
  data-testid="tableA"
>
  <table-table>
    <table-header>
      <column-header>
        ColA
      </column-header>
      <column-header>
        ColB
      </column-header>
    </table-header>
    <table-item>
      <column-text>
        a1
      </column-text>
      <column-set>
        <div>
          c1
        </div>
      </column-set>
      <table-buttons>
        <button
          data-testid="buttonEdit1"
        >
          Edit
        </button>
        <button
          data-testid="buttonDel1"
        >
          Del
        </button>
      </table-buttons>
    </table-item>
    <table-item>
      <column-text>
        a2
      </column-text>
      <column-set>
        -
      </column-set>
      <table-buttons>
        <button
          data-testid="buttonEdit2"
        >
          Edit
        </button>
        <button
          data-testid="buttonDel2"
        >
          Del
        </button>
      </table-buttons>
    </table-item>
  </table-table>
  <button
    data-testid="buttonAdd"
  >
    Add
  </button>
  <button
    data-testid="buttonClear"
  >
    Confirm
  </button>
</div-page>
`;

exports[`Table confirms & clears list 3`] = `
<div-page
  data-testid="tableA"
>
  <table-table>
    <table-header>
      <column-header>
        ColA
      </column-header>
      <column-header>
        ColB
      </column-header>
    </table-header>
  </table-table>
  <button
    data-testid="buttonAdd"
  >
    Add
  </button>
  <button
    data-testid="buttonClear"
  >
    Clear
  </button>
</div-page>
`;

exports[`Table enables a slot 1`] = `
<div-page
  data-testid="tableA"
>
  <table-table>
    <table-header>
      <column-header>
        ColA
      </column-header>
      <column-header>
        ColB
      </column-header>
    </table-header>
    <table-item>
      <column-text>
        a1
      </column-text>
      <column-set>
        <div>
          c1
        </div>
      </column-set>
      <table-buttons>
        <button
          data-testid="buttonEdit1"
        >
          Edit
        </button>
        <button
          data-testid="buttonDel1"
        >
          Del
        </button>
      </table-buttons>
    </table-item>
  </table-table>
  <button
    data-testid="buttonAdd"
  >
    Add
  </button>
  <button
    data-testid="buttonClear"
  >
    Clear
  </button>
</div-page>
`;

exports[`Table enters row into list 1`] = `
<div-page
  data-testid="tableA"
>
  <table-table>
    <table-header>
      <column-header>
        ColA
      </column-header>
      <column-header>
        ColB
      </column-header>
      <column-header>
        ColC
      </column-header>
    </table-header>
    <table-item>
      <column-text>
        a
      </column-text>
      <column-set>
        -
      </column-set>
      <column-text>
        2000/01/01
      </column-text>
      <table-buttons>
        <button
          data-testid="buttonEdit1"
        >
          Edit
        </button>
        <button
          data-testid="buttonDel1"
        >
          Del
        </button>
      </table-buttons>
    </table-item>
  </table-table>
  <button
    data-testid="buttonAdd"
  >
    Add
  </button>
  <button
    data-testid="buttonClear"
  >
    Clear
  </button>
</div-page>
`;

exports[`Table enters row into list 2`] = `
<div-page
  data-testid="tableA"
>
  <table-table>
    <table-header>
      <column-header>
        ColA
      </column-header>
      <column-header>
        ColB
      </column-header>
      <column-header>
        ColC
      </column-header>
    </table-header>
    <table-item>
      <column-text>
        <input
          data-testid="inputColA"
          value="a1"
        />
      </column-text>
      <column-set>
        <div>
          <input
            data-testid="inputColB1"
            type="checkbox"
          />
           
          c1
        </div>
      </column-set>
      <column-text>
        <input
          data-testid="inputColC"
          style="background: red;"
          value="2022-01-04"
        />
      </column-text>
      <table-buttons>
        <button
          data-testid="buttonSave"
        >
          Save
        </button>
        <button
          data-testid="buttonCancel"
        >
          Cancel
        </button>
      </table-buttons>
    </table-item>
  </table-table>
  <button
    data-testid="buttonAdd"
  >
    Add
  </button>
  <button
    data-testid="buttonClear"
  >
    Clear
  </button>
</div-page>
`;

exports[`Table enters row into list 3`] = `
<div-page
  data-testid="tableA"
>
  <table-table>
    <table-header>
      <column-header>
        ColA
      </column-header>
      <column-header>
        ColB
      </column-header>
      <column-header>
        ColC
      </column-header>
    </table-header>
    <table-item>
      <column-text>
        a1
      </column-text>
      <column-set>
        -
      </column-set>
      <column-text>
        2022/01/04
      </column-text>
      <table-buttons>
        <button
          data-testid="buttonEdit1"
        >
          Edit
        </button>
        <button
          data-testid="buttonDel1"
        >
          Del
        </button>
      </table-buttons>
    </table-item>
  </table-table>
  <button
    data-testid="buttonAdd"
  >
    Add
  </button>
  <button
    data-testid="buttonClear"
  >
    Clear
  </button>
</div-page>
`;

exports[`Table inlineHeaders 1`] = `
<div-page
  data-testid="tableA"
>
  <table-table>
    <table-item>
      <table-row
        style="display: block;"
      >
        <column-header>
          ColA
        </column-header>
        <div>
          <column-text>
            a1
          </column-text>
        </div>
      </table-row>
      <table-row
        style="display: block;"
      >
        <column-header>
          ColB
        </column-header>
        <div>
          <column-set>
            <div>
              c1
            </div>
          </column-set>
        </div>
      </table-row>
      <table-buttons>
        <button
          data-testid="buttonEdit1"
        >
          Edit
        </button>
        <button
          data-testid="buttonDel1"
        >
          Del
        </button>
      </table-buttons>
    </table-item>
  </table-table>
  <p>
    1
     - 
    1
     of 
    2
  </p>
  <button
    data-testid="buttonPrev"
  >
    Prev
  </button>
  <button
    data-testid="buttonNext"
  >
    Next
  </button>
  <button
    data-testid="buttonAdd"
  >
    Add
  </button>
  <button
    data-testid="buttonClear"
  >
    Clear
  </button>
</div-page>
`;

exports[`Table pages 1`] = `
<div-page
  data-testid="tableA"
>
  <table-table>
    <table-header>
      <column-header>
        ColA
      </column-header>
      <column-header>
        ColB
      </column-header>
    </table-header>
    <table-item>
      <column-text>
        a1
      </column-text>
      <column-set>
        <div>
          c1
        </div>
      </column-set>
      <table-buttons>
        <button
          data-testid="buttonEdit1"
        >
          Edit
        </button>
        <button
          data-testid="buttonDel1"
        >
          Del
        </button>
      </table-buttons>
    </table-item>
  </table-table>
  <p>
    1
     - 
    1
     of 
    2
  </p>
  <button
    data-testid="buttonPrev"
  >
    Prev
  </button>
  <button
    data-testid="buttonNext"
  >
    Next
  </button>
  <button
    data-testid="buttonAdd"
  >
    Add
  </button>
  <button
    data-testid="buttonClear"
  >
    Clear
  </button>
</div-page>
`;

exports[`Table pages next prev 1`] = `
<div-page
  data-testid="tableA"
>
  <table-table>
    <table-header>
      <column-header>
        ColA
      </column-header>
      <column-header>
        ColB
      </column-header>
    </table-header>
    <table-item>
      <column-text>
        a1
      </column-text>
      <column-set>
        <div>
          c1
        </div>
      </column-set>
      <table-buttons>
        <button
          data-testid="buttonEdit1"
        >
          Edit
        </button>
        <button
          data-testid="buttonDel1"
        >
          Del
        </button>
      </table-buttons>
    </table-item>
  </table-table>
  <p>
    1
     - 
    1
     of 
    2
  </p>
  <button
    data-testid="buttonPrev"
  >
    Prev
  </button>
  <button
    data-testid="buttonNext"
  >
    Next
  </button>
  <button
    data-testid="buttonAdd"
  >
    Add
  </button>
  <button
    data-testid="buttonClear"
  >
    Clear
  </button>
</div-page>
`;

exports[`Table pages next prev 2`] = `
<div-page
  data-testid="tableA"
>
  <table-table>
    <table-header>
      <column-header>
        ColA
      </column-header>
      <column-header>
        ColB
      </column-header>
    </table-header>
    <table-item>
      <column-text>
        a2
      </column-text>
      <column-set>
        -
      </column-set>
      <table-buttons>
        <button
          data-testid="buttonEdit2"
        >
          Edit
        </button>
        <button
          data-testid="buttonDel2"
        >
          Del
        </button>
      </table-buttons>
    </table-item>
  </table-table>
  <p>
    2
     - 
    2
     of 
    2
  </p>
  <button
    data-testid="buttonPrev"
  >
    Prev
  </button>
  <button
    data-testid="buttonNext"
  >
    Next
  </button>
  <button
    data-testid="buttonAdd"
  >
    Add
  </button>
  <button
    data-testid="buttonClear"
  >
    Clear
  </button>
</div-page>
`;

exports[`Table pages next prev 3`] = `
<div-page
  data-testid="tableA"
>
  <table-table>
    <table-header>
      <column-header>
        ColA
      </column-header>
      <column-header>
        ColB
      </column-header>
    </table-header>
    <table-item>
      <column-text>
        a2
      </column-text>
      <column-set>
        -
      </column-set>
      <table-buttons>
        <button
          data-testid="buttonEdit2"
        >
          Edit
        </button>
        <button
          data-testid="buttonDel2"
        >
          Del
        </button>
      </table-buttons>
    </table-item>
  </table-table>
  <p>
    2
     - 
    2
     of 
    2
  </p>
  <button
    data-testid="buttonPrev"
  >
    Prev
  </button>
  <button
    data-testid="buttonNext"
  >
    Next
  </button>
  <button
    data-testid="buttonAdd"
  >
    Add
  </button>
  <button
    data-testid="buttonClear"
  >
    Clear
  </button>
</div-page>
`;

exports[`Table pages next prev 4`] = `
<div-page
  data-testid="tableA"
>
  <table-table>
    <table-header>
      <column-header>
        ColA
      </column-header>
      <column-header>
        ColB
      </column-header>
    </table-header>
    <table-item>
      <column-text>
        a1
      </column-text>
      <column-set>
        <div>
          c1
        </div>
      </column-set>
      <table-buttons>
        <button
          data-testid="buttonEdit1"
        >
          Edit
        </button>
        <button
          data-testid="buttonDel1"
        >
          Del
        </button>
      </table-buttons>
    </table-item>
  </table-table>
  <p>
    1
     - 
    1
     of 
    2
  </p>
  <button
    data-testid="buttonPrev"
  >
    Prev
  </button>
  <button
    data-testid="buttonNext"
  >
    Next
  </button>
  <button
    data-testid="buttonAdd"
  >
    Add
  </button>
  <button
    data-testid="buttonClear"
  >
    Clear
  </button>
</div-page>
`;
